mk/nomagic branch as discussed

matthiask opened the pull request:
Lots of changes here. If reviewing gets too hard (or if you want to review smaller pieces) I'll happily split up the changes. 

(If that's not necessary I wont.)

craigds commented:
It occurred to me that `TreeManager.disable_mptt_updates` doesn't really add any magic; it just turns off recalculation of fields. Perhaps we could keep that even though we're dropping `delay_mptt_updates`

I'll ask the mailing list for further feedback on the changes. Thanks again

matthiask commented:
I've cleaned up the PR a bit, and removed the addition of bulk_load to keep the diff clean(er): (https://github.com/matthiask/django-mptt/tree/mk/nomagic-bulk-load)

matthiask commented:
While chasing down a surprising (at least in the beginning) crash I found a behavior change in the new code which must be documented, and I'm not yet clear on how to resolve all edge cases best.

The current code in django-mptt does NOT run `model.save()` within `TreeManager.move_node()`. The new code obviously has to run `model.save()` because the tree maintenance operations never touch the `parent` field, a consequence of the fix for https://github.com/django-mptt/django-mptt/issues/466.

Completely obvious in hindsight, but at first I was dumbfounded (we had crashes in [feincms3](https://github.com/matthiask/feincms3)'s pages administration and a [reproducible test failure](https://travis-ci.org/matthiask/feincms3/jobs/149103192) — more fixes forthcoming)

matthiask commented:
I have a few doubts regarding transaction isolation levels and reading MPTT values from the database. Not sure whether this approach is even viable. We should be alright with READ COMMITTED, but that's not the default on MySQL. 

(OTOH who even cares about correctness on MySQL. Sorry, could not resist.)

auvipy commented:
can this move forward?

orf commented on mptt/admin.py:
```diff
@@ -225,13 +234,14 @@ def _move_node(self, request):
         except IntegrityError as e:
             self.message_user(request, _('Database error: %s') % e)
-            raise
+            return http.HttpResponse('FAIL, invalid move.')
```
These should all have a 400 status code to go with them IMO.

orf commented on mptt/managers.py:
```diff
@@ -53,51 +52,40 @@
+    @contextlib.contextmanager
+    def disable_mptt_updates(self):
+        # Preserve API
+        yield
```
Maybe raise a deprecation warning here? It seems senseless to keep them around forever if they are defunct.

matthiask replied:
Yes.

orf commented on mptt/managers.py:
```diff
@@ -794,46 +553,34 @@ def _inter_tree_move_and_close_gap(
 
-        opts = self.model._meta
         inter_tree_move_query = """
         UPDATE %(table)s
```
I'm no expert on the internals of django-mptt, but why is raw sql needed here? Could you do:

```python
Model.objects.update(
    level=Case(When(lft__gte=x, then=F('level') - ....))),
    tree_id=Case(When(lft__gte=y...
)
```

Using the Case/When classes built into Django?

matthiask replied:
The reason is simple: The code is a bit older than Django's support for Case/When :)

Case/When should work too.

orf commented on mptt/managers.py:
```diff
@@ -952,62 +695,55 @@ def _make_sibling_of_root_node(self, node, target, position):
-            cursor.execute(space_query, [target, size, target, size, tree_id,
-                                         target, target])
+        space_query = """
+        UPDATE %(table)s
```
Same for this?

orf commented on mptt/managers.py:
```diff
@@ -1123,41 +854,30 @@ def _move_child_within_tree(self, node, target, position):
-                ELSE %(parent)s END
-        WHERE %(tree_id)s = %%s""" % {
+        SET level = CASE
+                WHEN lft >= %%s AND lft <= %%s
```
And this?

matthiask commented:
I'm now quite sure that this PR does not work with transaction isolation levels SERIALIZABLE and REPEATABLE READ. 

I read initial values, send a big UPDATE over the wire and read the same row again in the same transaction. We still had problems with MPTT attributes getting out of sync with the code in this PR. So maybe there are some good ideas in here, but I wouldn't recommend merging this as it is.

auvipy commented:
certainly this will need more works

orf commented:
Perhaps we should split it up into more manageable chunks? Less magic is certainly a good thing to strive for, there seems to be a fair bit here that can be made into separate MRs and reviewed/merged independently?

craigds commented:
Needs a bit of work put into bringing this up to date with master. Some commits are no longer relevant and others need a bit of refactoring.

I've started rebasing some of the commits here onto #567. That one will stay up-to-date with any master changes until it's merged.

I've started with the initial stable things that are obvious and I understand well, and I'll create some individual pull requests for the rest.
